File: src/Finos.Fdc3.Backplane.Client.JS/src/service/ClientMiddleware.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2021 FINOS FDC3 contributors - see NOTICE file
 */

import { BackplaneTransport } from '../transport/BackplaneTransport';
import { Logger } from 'ts-log';
import { Context, ContextHandler, Listener } from '@finos/fdc3';
import {
	buildBroadcastEnvelop,
	getBroadcastContextMapKey,
	onHubConnection,
	unsubscribeUtil,
} from '../utility/ClientMiddlewareUtils';
import { HubConnection } from '@microsoft/signalr';
import { AppIdentifier } from '../DTO/receive/MessageEnvelope';
import { BackplaneDiscoveryServiceClient } from '../discovery/BackplaneDiscoveryServiceClient';
import { ChannelsRepository } from '../repository/ChannelsRepository';
import { Fdc3Channel } from '../broadcast/channels/ChannelClient';

export class ClientMiddleware {
	private backplaneTransport: BackplaneTransport;
	private contextHandlers: Map<string, ContextHandler[]>;
	private logger: Logger;
	private appIdentifier: AppIdentifier;
	private channelsRepository: ChannelsRepository;

	constructor(appIdentifier: AppIdentifier, logger: Logger) {
		this.backplaneTransport = new BackplaneTransport(new BackplaneDiscoveryServiceClient(logger), logger);
		this.channelsRepository = new ChannelsRepository(this.backplaneTransport, this, logger);
		this.appIdentifier = appIdentifier;
		this.contextHandlers = new Map<string, ContextHandler[]>();
		this.backplaneTransport.registerOnConnect(() =>
			onHubConnection(this.logger, this.backplaneTransport, this.contextHandlers)
		);
		this.logger = logger;
	}

	getSystemChannels(): Fdc3Channel[] {
		return this.channelsRepository.getSystemChannels();
	}

	async initializeAsync(retryCount: number, retryIntervalInMs: number): Promise<void> {
		await this.backplaneTransport.initializeAsync(retryCount, retryIntervalInMs);
		await this.channelsRepository.InitializeAsync(retryCount, retryIntervalInMs);
	}

	broadcastContext(context: Context, channelId: string): void {
		let hubConn: HubConnection | undefined = this.backplaneTransport.getHubConnection();
		if (hubConn == undefined) throw Error('Invalid operation: Transport not initialized');
		hubConn.invoke(
			'Broadcast',
			JSON.parse(buildBroadcastEnvelop(this.appIdentifier, context, channelId, this.logger)),
			false
		);
	}

	addContextListener(contextType: string | undefined, handler: ContextHandler, channelId: string): Listener {
		this.logger.info(`Registering contextType: ${contextType} for listening from backplane`);
		let key: string = getBroadcastContextMapKey(contextType, channelId);
		if (this.contextHandlers.has(key)) {
			this.logger.info(`ContextType already found. Registering handler..`);
			this.contextHandlers.get(key)?.push(handler);
			this.logger.info(`Registered handler successfully: ${contextType}`);
		} else {
			this.contextHandlers.set(key, [handler]);
			this.logger.info(`Registered contextType and handler`);
		}

		return {
			unsubscribe: (): void => {
				this.logger.info(`Unsubscribe called for contextType: ${key}`);
				let item: ContextHandler[] | undefined = this.contextHandlers.get(key);
				unsubscribeUtil(this.logger, item, 'contextType', key, handler);
			},
		};
	}
}

File: src/Finos.Fdc3.Backplane.Client/Middleware/ClientMiddleware.cs
/**
	* SPDX-License-Identifier: Apache-2.0
	* Copyright 2021 FINOS FDC3 contributors - see NOTICE file
	*/


using Finos.Fdc3.Backplane.Client.API;
using Finos.Fdc3.Backplane.Client.Broadcast.Channels;
using Finos.Fdc3.Backplane.Client.Concurrency;
using Finos.Fdc3.Backplane.Client.Transport;
using Finos.Fdc3.Backplane.DTO.Envelope;
using Finos.Fdc3.Backplane.DTO.Envelope.Receive;
using Finos.Fdc3.Backplane.DTO.Envelope.Send;
using Finos.Fdc3.Backplane.DTO.FDC3;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reactive.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

[assembly: InternalsVisibleTo("Finos.Fdc3.Backplane.Client.Test")]
namespace Finos.Fdc3.Backplane.Client.Middleware
{
    internal class ClientMiddleware : IClientMiddleware
    {
        private readonly IBackplaneTransport _desktopAgentTransport;
        private readonly ILogger<IClientMiddleware> _logger;
        private readonly ISchedulerProvider _schedulerProvider;
        private AppIdentifier _appMetadata;

        public IObservable<ConnectionState> ConnectionStateStream => _desktopAgentTransport.ConnectionStateStream;

        public ClientMiddleware(ILogger<IClientMiddleware> logger,
            IBackplaneTransport desktopAgentTransport,
            ISchedulerProvider schedulerProvider)
        {
            _schedulerProvider = schedulerProvider;
            _desktopAgentTransport = desktopAgentTransport;
            _logger = logger;
        }


        public async Task<IListener> AddContextListenerAsync(string contextType, Action<IContext> handler, string channelId, CancellationToken ct = default)
        {
            string subscribedContextType = contextType;
            Action<IContext> handlerLocal = handler;
            string channelIdSubscribedLocal = channelId;
            _logger.LogInformation($"Adding context listener for context type: {contextType}");
            IDisposable disposable = _desktopAgentTransport.ReceiveDataStream.SubscribeOn(_schedulerProvider.TaskPool).Subscribe(messageEnvelope =>
            {
                try
                {
                    _logger.LogDebug($"Receive Broadcast handler: Received message from backplane with message {JsonConvert.SerializeObject(messageEnvelope)}");
                    if (messageEnvelope.ActionType != Fdc3Action.Broadcast)
                    {
                        return;
                    }

                    _logger.LogInformation($"Receive Broadcast handler: Received message envelope from backplane with id from {messageEnvelope.Meta.Source.AppId} with messageId: {messageEnvelope.Meta.UniqueMessageId}");
                    string json = messageEnvelope.Payload.Context.ToString();
                    Context receivedContext = new Context(messageEnvelope.Payload.Context);

                    // null context type indicates listen all context type
                    if (subscribedContextType == null || subscribedContextType == receivedContext.Type)
                    {
                        if (channelIdSubscribedLocal == messageEnvelope.Payload.ChannelId)
                        {
                            handler(new Context(messageEnvelope.Payload.Context));
                            _logger.LogDebug($"Handler invoked for received context envelope: {JsonConvert.SerializeObject(messageEnvelope)}");
                        }
                        else
                        {
                            _logger.LogDebug($"Not handling contextType: {receivedContext.Type} as it has different channel associated than subscribed: {channelIdSubscribedLocal}");
                        }
                    }
                    else
                    {
                        _logger.LogDebug($"Not handling contextType: {subscribedContextType} as it did not match subscribed context type: {subscribedContextType}");
                    }
                }
                catch (Exception ex)
                {
                    _logger.LogError($"Failed to handle broadcast context received from backplane: {ex}");
                }

            });
            return await Task.FromResult(new Listener(async (token) =>
            {
                disposable.Dispose();
                await Task.CompletedTask;
                _logger.LogInformation($"Unsubscribed from context type: {subscribedContextType}");
            }));
        }

        public async Task BroadcastAsync(IContext context, string channelId, CancellationToken ct = default)
        {
            BroadcastContextEnvelope messageEnvelope = new BroadcastContextEnvelope()
            {
                ChannelId = channelId,
                Context = (Context)context,
                Metadata = new EnvelopeMetadata() { Source = _appMetadata, UniqueMessageId = Guid.NewGuid().ToString() }
            };
            _logger.LogInformation($"Broadcasting context with messageId {messageEnvelope.Metadata.UniqueMessageId} on channel: {channelId} to backplane");
            await _desktopAgentTransport.BroadcastAsync(messageEnvelope, ct);

        }

        public void Dispose()
        {
            _logger.LogInformation("Disposing..");
            _desktopAgentTransport.Dispose();
        }

        public async Task<IContext> GetCurrentContextAsync(string channelId, CancellationToken ct = default)
        {
            return await _desktopAgentTransport.GetCurrentContextAsync(channelId, ct);
        }

        public async Task<IEnumerable<IChannel>> GetSystemChannelsAsync(CancellationToken ct = default)
        {
            IEnumerable<Channel> channelsDto = await _desktopAgentTransport.GetSystemChannelsAsync(ct);
            IEnumerable<ChannelClient> systemChannels = channelsDto?.Select(x => new ChannelClient(this, x));
            if (!systemChannels.Any())
            {
                throw new InvalidOperationException("No system channels populated!");
            }

            return systemChannels;
        }

        public async Task InitializeAsync(AppIdentifier appMetadata, int retryCount, Func<int, TimeSpan> retryIntervalProvider, CancellationToken ct = default)
        {
            _appMetadata = appMetadata;
            await _desktopAgentTransport.InitializeConnectionAsync(retryCount, retryIntervalProvider, ct);
        }
    }
}
